src/backend/distributed/planner/distributed_planner.c

```diff
@@ -2321,6 +2321,76 @@ HasUnresolvedExternParamsWalker(Node *expression, ParamListInfo boundParams)
 }
 
 
+/*
+ * IsExternParamUsedInQuery returns true if the passed in paramId
+ * is used in the query, false otherwise.
+ */
+bool
+IsExternParamUsedInQuery(Node *expression, int *ParamIdAddr)
+{
+	if (expression == NULL)
+	{
+		return false;
+	}
+
+	if (IsA(expression, Param))
+	{
+		Param *param = (Param *) expression;
+		int paramId = param->paramid;
+
+		/* only care about user supplied parameters */
+		if (param->paramkind != PARAM_EXTERN)
+		{
+			return false;
+		}
+		if (paramId == *ParamIdAddr)
+		{
+			/* Found the parameter */
+			return true;
+		}
+
+		/* Continue searching */
+		return false;
+	}
+
+	/* keep traversing */
+	if (IsA(expression, Query))
+	{
+		return query_tree_walker((Query *) expression,
+								 IsExternParamUsedInQuery,
+								 ParamIdAddr,
+								 0);
+	}
+	else
+	{
+		return expression_tree_walker(expression,
+									  IsExternParamUsedInQuery,
+									  ParamIdAddr);
+	}
+}
+
+
+/*
+ * MarkUnreferencedExternParams marks parameter's type to zero if the
+ * parameter is not used in the query.
+ */
+void
+MarkUnreferencedExternParams(Node *expression, ParamListInfo boundParams)
+{
+	int parameterCount = boundParams->numParams;
+
+	for (int parameterIndex = 0; parameterIndex < parameterCount; parameterIndex++)
+	{
+		int paramId = parameterIndex + 1;
+
+		if (!IsExternParamUsedInQuery(expression, &paramId))
```

Comment:
this repeated walker seems slow, probably not something we can release without perf tests

Reply:
Trying to run our performance tests on CircleCI, but it's down, so, I tried to create a simple 10K executions of a prepared statement. With multiple runs, average seems to suggest no change, 393941ms (this PR) vs 453844ms (master), but the performance numbers on a personal machine in an uncontrolled environment are not reliable. I will update the PR as I get more numbers.

Reply:
![image](https://user-images.githubusercontent.com/44680808/164323540-658a30c7-70a9-420b-b0f2-f7a48bb975c7.png)

Reply:
![image](https://user-images.githubusercontent.com/44680808/164325334-c3c6b282-e4ab-45b4-9f16-5ef7b0337fff.png)

Reply:
From master
![image](https://user-images.githubusercontent.com/44680808/164337373-15abfcdc-006e-42fb-a05d-5d8e263b4fef.png)

Reply:
From Master second run
![image](https://user-images.githubusercontent.com/44680808/164371745-eb8fc7ea-b29b-4323-8525-88ef2871563e.png)

Reply:
**Master**
                        latency         tps
With xact:
default-pgbench         15.070          2122.832837
simple-update           12.495          2559.896604
select-only             0.832           38301.142324
**PR**
                        latency         tps
With xact:
default-pgbench         16.856          1897.531550
simple-update           12.903          2479.139232
select-only             1.068           29849.587196


**Master**
Without xact:
default-pgbench         26.745          1196.289699
simple-update           15.953          2005.279067
select-only             0.856           37215.900488
**PR**
Without xact:
default-pgbench         27.311          1171.501152
simple-update           15.327          2087.242627
select-only             1.075           29673.977062

Reply:
If these numbers are accurate, I see impact on select-only

Reply:
@marcocitus I've changed the code to walk the Query only once, can you please take a look.

Reply:
@marcoslot